File: source/go/memory_pool.go
package main

import "fmt"

// Fixed-size object pool: 8 slots of 8 bytes each
const slotSize = 8
const poolSlots = 8

var pool [poolSlots][slotSize]byte
var used [poolSlots]bool
var allocCount = 0

func poolAlloc() int {
	for i := 0; i < poolSlots; i++ {
		if !used[i] {
			used[i] = true
			allocCount++
			fmt.Printf("alloc slot %d\n", i)
			return i
		}
	}
	fmt.Println("pool exhausted!")
	return -1
}

func poolFree(slotIndex int) {
	if used[slotIndex] {
		used[slotIndex] = false
		allocCount--
		fmt.Printf("freed slot %d\n", slotIndex)
	}
}

func activeCount() int {
	count := 0
	for i := 0; i < poolSlots; i++ {
		if used[i] {
			count++
		}
	}
	return count
}

func main() {
	s0 := poolAlloc()
	s1 := poolAlloc()
	s2 := poolAlloc()
	s3 := poolAlloc()
	s4 := poolAlloc()

	// Write data (simulate ptr_write_i64)
	_ = s0
	_ = s1
	_ = s2
	_ = s3
	_ = s4

	fmt.Printf("active: %d\n", activeCount())

	poolFree(1)
	poolFree(3)

	fmt.Printf("active after free: %d\n", activeCount())

	poolAlloc()
	poolAlloc()
	poolAlloc()

	fmt.Printf("final active: %d\n", activeCount())

	fmt.Println("pool freed")
}

File: source/c/memory_pool.c
// gcc -o memory_pool memory_pool.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SLOT_SIZE 8
#define POOL_SLOTS 8

char pool_buf[SLOT_SIZE * POOL_SLOTS];
int used[POOL_SLOTS];
int alloc_count = 0;

void pool_init(void) {
    memset(used, 0, sizeof(used));
    alloc_count = 0;
}

void *pool_alloc(void) {
    for (int i = 0; i < POOL_SLOTS; i++) {
        if (!used[i]) {
            used[i] = 1;
            alloc_count++;
            printf("alloc slot %d\n", i);
            return pool_buf + i * SLOT_SIZE;
        }
    }
    printf("pool exhausted!\n");
    return NULL;
}

void pool_free(int slot_index) {
    if (used[slot_index]) {
        used[slot_index] = 0;
        alloc_count--;
        printf("freed slot %d\n", slot_index);
    }
}

int active_count(void) {
    int count = 0;
    for (int i = 0; i < POOL_SLOTS; i++) if (used[i]) count++;
    return count;
}

int main(void) {
    pool_init();
    long long *s0 = pool_alloc();
    long long *s1 = pool_alloc();
    long long *s2 = pool_alloc();
    long long *s3 = pool_alloc();
    long long *s4 = pool_alloc();

    *s0 = 100; *s1 = 200;

    printf("active: %d\n", active_count());

    pool_free(1); pool_free(3);

    printf("active after free: %d\n", active_count());

    pool_alloc(); pool_alloc(); pool_alloc();

    printf("final active: %d\n", active_count());
    printf("pool freed\n");
    return 0;
}
